Les principes cachés du test (#171)
sogilis/Blog: site/content/posts/2020-10-01-principes-caches-du-test.md

---
title: Les principes cachés du test
author: jb@sogilis.com & willy@sogilis.com
date: 2020-10-01
image: /img/2020-10-01-principes-caches-du-test.jpg
altimage: '"Ten" by Mark Bonica is licensed with CC BY 2.0. To view a copy of this license, visit https://creativecommons.org/licenses/by/2.0/'
categories:
  - DÉVELOPPEMENT
tags:
  - test
---

## De quoi allons-nous parler ?

> Dans cet article, nous parlerons de certains types de tests logiciels. Plus précisément, il sera question de tests automatisés (pas de tests manuels) portant sur une base de code maintenable et évolutive (pas de code jetable, de [POC](https://fr.wikipedia.org/wiki/Preuve_de_concept), etc.).

Author: Oui, ce n'est pas forcément super fluide.
Par contre, je pense qu'il ne faut pas laisser entendre que le scope correspond à tout test automatisé, sinon on met un pas sur le terrain du dogmatisme.

> Dans cet article, nous parlerons de certains types de tests logiciels. Plus précisément, il sera question de tests automatisés (pas de tests manuels) portant sur une base de code maintenable et évolutive (pas de code jetable, de [POC](https://fr.wikipedia.org/wiki/Preuve_de_concept), etc.).

Reviewer: oui je suis d'accord, je ne voulais pas dire de supprimer le reste :)  "nous parlerons de tests automatisés (pas de tests manuels)  portant sur une base de code maintenable et évolutive (pas de code jetable, de [POC](https://fr.wikipedia.org/wiki/Preuve_de_concept), etc.)."

> Dans cet article, nous parlerons de certains types de tests logiciels. Plus précisément, il sera question de tests automatisés (pas de tests manuels) portant sur une base de code maintenable et évolutive (pas de code jetable, de [POC](https://fr.wikipedia.org/wiki/Preuve_de_concept), etc.).

Author: Ah oui, je vois ! 👍

Lorsque l'on travaille avec ce type de tests, de nombreux problèmes / difficultés / frustrations peuvent apparaître. Heureusement, certains **principes de base** existent pour limiter cela. Certains appellent cela des « bonnes pratiques ».

D'après nos expériences chez Sogilis, certaines de ces pratiques sont déjà populaires et largement adoptées dans le métier du développement logiciel. On pourra citer :

- Chaque test ne vérifie qu'un seul cas fonctionnel
- Indépendance entre les tests
- Exécution rapide
- Automatisation du setup de l'environnement test
- Tests indépendants de l'OS
- Exécutables plusieurs fois d'affilée (pas d'effet de bord sur l'environnement de test)

> Dans cet article, nous allons parler de **stratégies moins connues**, mais qui permettent de gagner encore un peu plus en lisibilité, maintenabilité, évolutivité et **utilisabilité au quotidien**.

Reviewer 2: On dit déjà "dans cet article" un peu plus haut. On pourrait trouver une autre formulation.

Author: Que penses-tu de cela ?
```suggestion
Ici, nous aborderons des **stratégies moins connues**, mais qui permettent de gagner encore un peu plus en lisibilité, maintenabilité, évolutivité et **utilisabilité au quotidien**.
```

func TestHeroDAO_ResurrectAllKnights(t *testing.T) {
    // given
    sql.Register("mysql", &MySQLDriver{})
    dsn := fmt.Sprintf("%s:%s@tcp(%s)/%s", username, password, hostname, dbName)
    db, err := sql.Open("mysql", dsn)
    if err != nil {
        t.Logf("Error %s when opening DB\n", err)
        return
    }
    defer db.Close()

    db.Exec("TRUNCATE TABLE dead_hero")
    db.Exec("INSERT INTO dead_hero (name, allegiance, caste, dateOfDeath) VALUES ('Bohort', 'Gaunnes', 'knight', '25/10/1203')")

    // when
    kights := dao.NewHeroDAO(db).resurrectAllKnights()

    // then
    knightsContainsBohort := false
    for _, knight := range kights {
        if knight.name == "Bohort" {
            knightsContainsBohort = true
        }
    }
    if !knightsContainsBohort {
        t.Fatalf("Expected %s to contains 'Bohort'", knights)
    }
}
```

En réduisant le boilerplate a l'extrème, on peut arriver à ceci :

```go {linenos=inline}
func TestHeroDAO_ResurrectAllKnights(t *testing.T, heroDAO *HeroDAO) {
    // given
    dbtest.InsertDeadKnight("Bohort")

    // when
    kights := heroDAO.resurrectAllKnights()

    // then
    assertOneKnightWithName(t, knights, "name")
}
```

> - toute la partie connexion de la base de donnée a été déplacée en amont, pour être exécutée avant chaque test.

Author: Pourquoi retirer le tier ? Du moins pourquoi ici et pas sur les lignes suivantes ?

> - toute la partie connexion de la base de donnée a été déplacée en amont, pour être exécutée avant chaque test.

Reviewer: retirer le tiret? Je pense que c'est une erreur. J'ai édité mon commit.

- idem avec le `TRUNCATE` (ligne 12).
- ainsi que la création de `HeroDAO` (ligne 16), tous les tests de ce fichier ayant besoin du même `HeroDAO`.

De plus, notons que le boilerplate est souvent commun a plusieurs tests. Son extraction permet alors de le factoriser, facilitant sa maintenance, en minimisant la duplication. Attention toutefois à ne pas tomber dans l'excès en cherchant à factoriser trop de choses. L'élément important est d'extraire le boilerplate, sa factorisation n'est qu'une option, à étudier de façon opportuniste.

**Note :** nous avons observé que plus le périmètre du test est grand, plus ce boilerplate a tendance à être important.

### Limites

Certains outils, frameworks ou librairies peuvent limiter cette approche, obligeant à garder une certaine dose de boilerplate dans le code du test[^testcode].

## 2. Expliciter le cas de test complet {#explicite-test-case}

> Un nouveau développeur doit pouvoir comprendre rapidement le cas testé. Pourquoi ? Parce qu'un développeur passe beaucoup de temps a lire du code (notamment des tests), et le cas de test est finalement la première étape permettant de comprendre le code du test. C'est un peu comme n'importe quel sujet : on commence par assimiler les concepts globaux, puis on rentre de plus en plus dans les détails. Faire le chemin dans l'autre sens est plus difficile.

Reviewer 3: ```suggestion
Un nouveau développeur doit pouvoir comprendre rapidement le cas testé. Pourquoi ? Parce qu'un développeur passe beaucoup de temps à lire du code (notamment des tests), et le cas de test est finalement la première étape permettant de comprendre le code du test. C'est un peu comme n'importe quel sujet : on commence par assimiler les concepts globaux, puis on rentre de plus en plus dans les détails. Faire le chemin dans l'autre sens est plus difficile.
```

        .andExpect(jsonPath("count", Matchers.equalTo(7)));
}
```

D'où sort ce chiffre `7`, ligne 5 ? Pourquoi 7 et pas 42 ? Le problème ici est que le jeu de données est initialisé dans une méthode déclarée 500 lignes au-dessus, et appelée automatiquement par le framework de test.

Quelque chose comme cela aurait été plus limpide pour le lecteur :

```kotlin {linenos=inline}
@Test
fun `/knights returns dummy knights`() {
    repeat(7) {
        knightRepository.create(KnightFactory.dummy())
    }
    mockMvc.perform(get("/knights?dummy=true"))
        .andExpect(status().isOk())
        .andExpect(jsonPath("count", Matchers.equalTo(7)));
}
```

Autre exemple plus compliqué :

```javascript {linenos=inline}
describe('TournamentAPI', () => {
    beforeEach(() => axios.post.mockReset());

    it('.joust() returns winner', async () => {
        const loser = buildHero();
        const winner = mockWinner()
        const returnedWinner = fight(winner, loser)
        expect(returnedWinner).toEqual(winner);
    });
});
```

Ce test est assez confus :
- on ne sait pas trop ce que mock la fonction `mockWinner()` (ligne 6), ni ce que cette fonction retourne.

> - que veut-on tester ? `TournamentAPI.joust()` ou `fight()` ?

Reviewer 4: Je trouve cet exemple un peu flou. D'ou sort le TournamentAPI.joust()? Pourquoi mockWinner est pas bon et buildHero ok? Pourquoi tu ne dis rien du loser = buildHero? On pourrait penser buildLoser()?

Author: Tu as raison, l'exemple n'est pas génial.

Peut-être qu'il faut quelque chose de plus simple comme ceci ?

Autre exemple :

```javascript {linenos=inline}
 describe('Excalibur', () => {
     it('changeBearer() returns sparkling excalibur when bearer has destiny', async () => {
         const hero = buildHero();
         excalibur = buildExcalibur().changeBearer(hero)
         expect(excalibur.isSparkling).toBeTruthy();
     });
 });
 ```

En lisant le code, on comprend que n'importe quel hero peut faire étinceler excalibur, alors que seuls ceux qui ont une destiné devraient pouvoir. Pourtant le test fonctionne puisque `buildHero()` retourne un héro ayant une destiné :

 ```javascript
function buildHero() {
   return {
      id: 3,
      name: 'Perceval,
      hasDestiny: true,
      rank: 'Knight',
   };
}
 ```

Soit cette fonction `buildHero()` est mal nommée (et devrait s'appeler `buildHeroWithDetiny()`), soit il manque quelque chose dans le test (comme `hero.hasDestiny = true`).

Reviewer 5: la dernière version est mieux :+1:

Author: Exemple modifié

## 4. Réfléchir au ROI

Lorsque l'on commence à écrire un test, cela vaut le coup de se poser 10 secondes pour réfléchir au [ROI](https://fr.wikipedia.org/wiki/Retour_sur_investissement) de ce test.

En effet, au-delà de ce qu'apporte un test, **il a aussi un coût**, qu'il soit **immédiat** (temps d'écriture du test) ou **récurrent** (maintenance, rallongement de la durée d'exécution des tests, augmentation du coût de certains refactoring…). Tous ces facteurs ne sont pas à négliger car ils peuvent avoir des conséquences.

Bien sûr, ces différents facteurs **dépendent du contexte**. Par exemple, sur une application critique, ce surcoût sera peut-être largement amorti par la réduction du risque de défaillance et leurs conséquences en perte financière ou humaine…

En revanche, il n'est clairement pas facile d'évaluer tout cela rapidement, et encore moins avec un niveau de confiance élevé. L'expérience peut aider, mais cela passe nécessairement par la **prise de recul régulier** sur le travail des autres, mais aussi **son propre travail**.

Voici quelques axes de coûts pour aider à la réflexion :

> - plus un test est bas niveau[^testlevel], plus il y a de risque d'engendrer un surcoût lors d'un refactoring.

Author: From @Isabelledonne :
??? mériterait une petite explication ou un exemple ? pourquoi est-ce que c'est le cas ?

Author: Ca mériterait pas mal d'explications pour décrire ça je pense, idéalement avec des schémas, voir du code. Du coup, je ne sais pas trop comment traiter ça.

- supprimer cette ligne (c'est dommage, c'est un élément intéressant dans la réflexion je trouve)
- ajouter une note de base de page (je ne suis pas sûr de parvenir à expliquer ça clairement en un seul paragraphe)
- autre ?

Ca pourrait être le sujet d'un autre article de blog par contre… et ferrait alors pointer cette ligne sur le nouvel article.

Reviewer 6: La dernière solution (faire référence à un article à venir) me paraît parfaite.

Author: Done

- plus le test est haut niveau, plus il est coûteux à mettre en place.

> - il est plus coûteux de trouver la cause d'un test en échec s'il est de haut niveau que s'il est de bas niveau.

Author: Non, pour moi, ce n'est pas directement lié.

- **Créer une nouvelle fixture ?** Multiplier les fixtures alourdi la modification du modèle de données.

Les fixtures présentent un autre inconvénient majeur : elles masquent des informations qui devraient être dans le test (cf principe [Expliciter le cas de test complet](#explicite-test-case))).

Pour toutes ces raisons, il est préférable de privilégier la création de données à la volée à travers des functions claires et paramétrables (ex: [Object Mother et Builder en Java](/posts/2019-01-11-object-mother-builder-java/)).

**Note :** utilisées avec des tests transactionnels, les fixtures cependant peuvent grandement améliorer la rapidité d'exécution des tests (en chargeant les données en base une seule fois). À garder en tête lorsque le gain en rapidité d'exécution surpasse les inconvénients mentionnés ci-dessus.

## 7. Commenter en dernier recours, mais commenter

Si le contexte (techno utilisée, contraintes du projets…) ne permet pas d'avoir un code explicitant clairement et rapidement le cas de test, alors, et seulement alors, il ne faut pas hésiter à commenter le teste pour expliciter certains points.

> Certes les commentaires présentent de nombreux inconvénients, mais dans certaines situations bien particulières, les bénéfices peuvent l'emporter.

Author: Je pense qu'on peut l'accepter. En effet, si on veut éviter ce type de question, j'ai bien peur qu'un article de blog ne soit pas suffisant, tellement on aborde de concepts et tellement il y a de limite dans telle ou telle situation.

## Conclusion

Encore une fois, ces principes ne sont **pas des lois** et n'ont de sens que **dans un contexte donné**. Souvent, en fonction de la situation, il est préférable de faire des **compromis**, c'est l'une des difficultés de notre métier : adapter nos choix aux contraintes du moment.

Toujours est-il qu'en gardant en ligne de mire ces quelques principes, nous pouvons gagner en "utilisabilité" au quotidien.

## Définitions

> [^testcode]: le **code du test** est le code présent dans la fonction décrivant le test. On ne parle ici que du code appelé indirectement.

Author: From @Isabelledonne :
pas sûre de comprendre exactement ce qu'on veut dire par "appelé indirectement" ?

Author: Pour moi, lorsqu'une fonction A appelle une fonction B qui appelle une fonction C, alors A appelle C indirectement.

Peut-être quelque chose comme ça pour être plus clair ?
```suggestion
[^testcode]: le **code du test** est le code présent dans la fonction décrivant le test. On ne parle ici que du code appelé indirectement dans l'implémentation du test.
```

Reviewer 6: Avec l'exemple je comprends bien ce que tu entends par "appelé indirectement".
S'il n'y a que moi qui me suis posé la question, et que tous les autres ont tout de suite compris que ça correspondait à ceci : "lorsqu'une fonction A appelle une fonction B qui appelle une fonction C, alors A appelle C indirectement.", alors l'exemple est superflu.
Sinon, peut-être que l'exemple serait utile...

Author: J'ai du mal à trouver une formulation qui explique ça clairement (le mieux serait un exemple, mais pas trop la place dans ces footnotes). Et n'ayant qu'un seul retour sur ce point, je lâche l'affaire, mais à charge de revanche 😉.

> [^boilerplate]: le **boilerplate** est du code purement technique nécessaire au fonctionnement du test, mais ne portant aucune information sur le cas testé. Exemples : `TRUNCATE` de tables avant ou après le test, création d'utilisateur par défaut, instanciation d'une classe dont on teste une méthode…

Author: From @Isabelledonne :
Techniquement, la définition du boilerplate peut-être contestée (dans l'acceptation usuelle du terme, c'est plutôt du code répétitif que du code technique), mais on est d'accord que dans la pratique, dans les tests, les deux notions vont la plupart du temps recouvrir les même portions de code.

> boilerplate : se dit pour une fonctionnalité d'un programme informatique dont le code source est quasiment le même quel que soit le programme (boilerplate code)

Author: Tout a fait, peut-être peut-on ajouter ceci pour être moins sujet a débat?

```suggestion
[^boilerplate]: le mot **boilerplate** est utilisé ici pour décrire du code purement technique nécessaire au fonctionnement du test, mais ne portant aucune information sur le cas testé. Exemples : `TRUNCATE` de tables avant ou après le test, création d'utilisateur par défaut, instanciation d'une classe dont on teste une méthode…
```

Reviewer 6: Oui parfait, du coup on pose les conventions pour l'acceptation du terme "boilerplate" dans notre cas.
Ca me paraît une bonne solution.

---
title: Les principes cachés du test
author: jb@sogilis.com & willy@sogilis.com
date: 2020-10-01
image: /img/2020-10-01-principes-caches-du-test.jpg
altimage: '"Ten" by Mark Bonica is licensed with CC BY 2.0. To view a copy of this license, visit https://creativecommons.org/licenses/by/2.0/'
categories:
  - DÉVELOPPEMENT
tags:
  - test
---

## De quoi allons-nous parler ?

> Dans cet article, il est question de certains type de tests logiciels, ou plus précisément :
> 
> - de tests automatisés (pas de tests manuels)
> - base de code maintenable et évolutive (pas de code jetable, de [POC](https://fr.wikipedia.org/wiki/Preuve_de_concept)…)

Reviewer 7: ```suggestion
Dans cet article, nous parlerons de certains types de tests logiciels. Plus précisément, il sera question de tests automatisés (pas de tests manuels) portant sur une base de code maintenable et évolutive (pas de code jetable, de [POC](https://fr.wikipedia.org/wiki/Preuve_de_concept), etc.).
```

D'après nos expériences chez Sogilis, certaines de ces pratiques sont déjà populaires et largement adoptées dans le métier du développement logiciel. On pourra citer :

- Chaque test ne vérifie qu'un seul cas fonctionnel
- Indépendance entre les tests
- Exécution rapide
- Automatisation du setup de l'environnement test
- Tests indépendants de l'OS
- Exécutables plusieurs fois d'affilée (pas d'effet de bord sur l'environnement de test)

Dans cet article, nous allons parler de **stratégies moins connues**, mais qui permettent de gagner encore un peu plus en lisibilité, maintenabilité, évolutivité et **utilisabilité au quotidien**.

**Mise en garde :** Les règles présentées dans cet article sont issues de nos humbles expériences. Elles nous ont aidé, à plusieurs reprises, à développer des bases de tests maintenables, évolutives, et agréables pour les développeurs. Cependant, ce ne sont **pas des règles absolues** et doivent être adaptées, le cas échant, en fonction de la situation. Comme pour toute pratique en développement logiciel, ne soyez pas dogmatique !

## 1. Réduire le boilerplate[^boilerplate]

> L'idée derrière ce principe est de faciliter au maximum la compréhension du code du test[^testcode], par exemple pour la maintenance future par un développeur tiers. En effet, lorsqu'on est amené à faire évoluer un test existant, il est important de **comprendre le cas de test rapidement**. Et si ce code est **noyée au milieu de détails techniques**, cette tâche devient fastidieuse.

Reviewer 7: ```suggestion
L'idée derrière ce principe est de faciliter au maximum la compréhension du code du test[^testcode], par exemple pour la maintenance future par un développeur tiers. En effet, lorsqu'on est amené à faire évoluer un test existant, il est important de **comprendre le cas de test rapidement**. Et si ce code est **noyé au milieu de détails techniques**, cette tâche devient fastidieuse.
```

**Note :** nous avons observé que plus le périmètre du test est grand, plus ce boilerplate a tendance à être important.

### Limites

Certains outils, frameworks ou librairies peuvent limiter cette approche, obligeant à garder une certaine dose de boilerplate dans le code du test[^testcode].

## 2. Expliciter le cas de test complet {#explicite-test-case}

Un nouveau développeur doit pouvoir comprendre rapidement le cas testé. Pourquoi ? Parce qu'un développeur passe beaucoup de temps a lire du code (notamment des tests), et le cas de test est finalement la première étape permettant de comprendre le code du test. C'est un peu comme n'importe quel sujet : on commence par assimiler les concepts globaux, puis on rentre de plus en plus dans les détails. Faire le chemin dans l'autre sens est plus difficile.

> Et pour comprendre ce cas de test rapidement, il n'y a pas 36 solutions : il doit être décrit entièrement dans le code du test[^testcode]. Pas d'indirection, pas de logique caché dans un autre fichier, etc. Après avoir lu ce code de test, le lecteur doit être capable de répondre aux questions suivantes :

Reviewer 7: ```suggestion
Et pour comprendre ce cas de test rapidement, il n'y a pas 36 solutions : il doit être décrit entièrement dans le code du test[^testcode]. Pas d'indirection, pas de logique caché dans un autre fichier, etc. Après avoir lu ce code de test, le développeur doit être capable de répondre aux questions suivantes :
```

- que veut-on tester ? `TournamentAPI.joust()` ou `fight()` ?

L'intention derrière ce test était peut-être ceci :

```javascript
describe('TournamentAPI', () => {
    beforeEach(() => axios.post.mockReset());

    it('.joust() returns winner', async () => {
        axios.post.mockResolvedValue({ winnerId: 1 });
        const winner = buildHero();
        const loser = buildHero();
        const id = TournamentAPI.joust(winner, loser)
        expect(id).toEqual(1);
    });
});
```

Concrètement, les fonctions `mockWinner()` et `fight()` masquaient plusieurs choses. Il fallait aller lire le contenu de ces fonctions pour comprendre quel était l'objet du test.

**Note :** ce principe allié au précédent n'est finalement que l'expression du principe ["Code at Wrong Level of Abstraction"](https://moderatemisbehaviour.github.io/clean-code-smells-and-heuristics/general/g6-code-at-wrong-level-of-abstraction.html) exprimé dans *Clean Code*.

## 3. Minimiser le jeu de données

Le jeu de données décrit les conditions initiales du test. Plus ces **données sont nombreuses**, plus il est **difficile de comprendre** ou **de debugger le test**.

> C'est pourquoi, autant que faire se peut, il est préférable de limiter au maximum le jeu de test utilisé. Les jeux de données communs a plusieurs tests (généralement utilisé avec les fixtures de test[^testfixture]) sont à éviter, mais nous en reparlerons plus tard.

Reviewer 7: ```suggestion
C'est pourquoi, autant que faire se peut, il est préférable de limiter au maximum le jeu de test utilisé. Les jeux de données communs a plusieurs tests (généralement utilisés avec les fixtures de test[^testfixture]) sont à éviter, mais nous en reparlerons plus tard.
```

## 4. Réfléchir au ROI

Lorsque l'on commence à écrire un test, cela vaut le coup de se poser 10 secondes pour réfléchir au [ROI](https://fr.wikipedia.org/wiki/Retour_sur_investissement) de ce test.

En effet, au-delà de ce qu'apporte un test, **il a aussi un coût**, qu'il soit **immédiat** (temps d'écriture du test) ou **récurrent** (maintenance, rallongement de la durée d'exécution des tests, augmentation du coût de certains refactoring…). Tous ces facteurs ne sont pas à négliger car ils peuvent avoir des conséquences.

> Bien sûr, ces différents facteurs **dépendent du contexte**. Par exemple, sur une application critique, ce surcoût sera peut-être largement amorti par la réduction du risque de défaillance et leurs conséquences en perte financière ou humaines…

Reviewer 7: ```suggestion
Bien sûr, ces différents facteurs **dépendent du contexte**. Par exemple, sur une application critique, ce surcoût sera peut-être largement amorti par la réduction du risque de défaillance et leurs conséquences en perte financière ou humaine…
```

En revanche, il n'est clairement pas facile d'évaluer tout cela rapidement, et encore moins avec un niveau de confiance élevé. L'expérience peut aider, mais cela passe nécessairement par la **prise de recul régulier** sur le travail des autres, mais aussi **son propre travail**.

> Voici quelques axes de coûts pour aider à la réflexion :

Reviewer 7: Axes de coûts ? Est-il possible de reformuler ?
-> Voici quelques exemples concrets rencontrés en pratiques ?

Author: Que penses-tu de ceci ?
```suggestion
Voici quelques exemples concrets rencontrés :
```

- plus un test est bas niveau[^testlevel], plus il y a de risque d'engendrer un surcoût lors d'un refactoring.
- plus le test est haut niveau, plus il est coûteux à mettre en place.
- il est plus coûteux de trouver la cause d'un test en échec s'il est de haut niveau que s'il est de bas niveau.
- l'évolution du modèle de données peut nécessiter la modification de beaucoup de tests (exemple : l'ajout d'un attribut obligatoire dans une entité nécessite d'adapter tous les tests qui construisent une instance de cette entité).

> Heureusement, des stratégies (architectures, frameworks, etc.) existent pour limiter ce surcoût, mais toutes ne sont pas disponibles quelque soit le contexte.

Reviewer 7: ```suggestion
Heureusement, des stratégies (architectures, frameworks, etc.) existent pour limiter ce surcoût, mais toutes ne sont pas disponibles dans tous les contextes.
```

## 5. Assertions limpides

Il peut arriver que l'intention derrière certaines assertions ne soit pas exprimée très clairement, comme dans cet exemple :

```java
@Test
void parseThrowsParseException() {
    given(pets.findPetTypes()).willReturn(makePetTypes());
    try {
        petTypeFormatter.parse("Fish", Locale.ENGLISH);
        Assertions.fail();
    } catch (ParseException e) {
        String msg = e.getCause().getMessage();
        Matcher matcher = Pattern.compile("type not found:.*").matcher(msg);
        Assertions.assertTrue(matcher.find());
    }
}
```

> Comprendre quelle était l'intention demande un soupçon d'investigation, alors qu'il existe souvent des solutions peut coûteuses pour améliorer la situation.

Reviewer 7: ```suggestion
Comprendre quelle était l'intention demande un soupçon d'investigation, alors qu'il existe souvent des solutions peu coûteuses pour améliorer la situation.
```

> En voici un premier exemple naïf où l'on a créé préalablement une méthode `assertMessageMatches()` :

Reviewer 7: ```suggestion
Voici un premier exemple naïf où l'on a créé préalablement une méthode `assertMessageMatches()` :
```

```java
@Test
void parseThrowsParseException() {
    given(pets.findPetTypes()).willReturn(makePetTypes());
    try {
        petTypeFormatter.parse("Fish", Locale.ENGLISH);
        Assertions.fail();
    } catch (ParseException e) {
        assertMessageMatches(e.getCause(), "type not found:.*");
    }
}
```

… et un second, avec l'utilisation de la librairie [AssertJ](https://assertj.github.io/doc/):

```java
@Test
void parseThrowsParseException() {
    given(this.pets.findPetTypes()).willReturn(makePetTypes());
    assertThatThrownBy(() -> petTypeFormatter.parse("Fish", Locale.ENGLISH))
        .isInstanceOf(ParseException.class)
        .extracting(Throwable::getCause)
        .hasMessageMatching("type not found:.*");
}
```

Le surcoût engendré par cette étape de refactoring peut être amorti :
- lorsque la même assertion doit être faite **fréquemment**
- avec l'utilisation d'une lib tierce, et si cette lib tierce est **bien testée** (il y a alors moins de risque de bug)
- l'utilisation d'une **lib tierce largement employée** dans l'écosystème permet d'améliorer la productivité des nouveaux dev (vu qu'il y a de bonnes chances qu'ils connaissent cette lib)

> **Note :** l'utilisation d'assertions personnalisées (cf exemple naïf) est une option a ne pas négliger

Reviewer 7: ```suggestion
**Note :** l'utilisation d'assertions personnalisées (cf. exemple naïf) est une option a ne pas négliger.
```

## 6. Éviter les fixtures de test

Lors de l'écriture d'un test nécessitant un état particulier en **base de donnée**, l'utilisation de fixtures de test[^testfixture] peut se faire de différentes manières, mais toutes posent problème :

> - **Utiliser une fixture existante ?** Le jeu de données sera nécessairement plus important que nécessaire, puisque partagé entre plusieurs tests qui requièrent chacun des données différentes. En cas de test en échec, l'investigation sera plus difficile, les données intéressantes étant polluées.

Reviewer 7: double emploi de nécessaire ici à éviter

Author: 1er nécessaire remplacé par "alors".

> - **Adapter une fixture existante ?** On prend alors le risque de casser un autre test qui utilise cette fixture car on ne sait pas quel élément de la fixture est importante pour quel test.

Reviewer 7: ```suggestion
- **Adapter une fixture existante ?** On prend alors le risque de casser un autre test qui utilise cette fixture car on ne sait pas quel élément de la fixture est important pour quel test.
```

> - **Créer une nouvelle fixture ?** Multipliant les fixtures alourdi la modification du modèle de données.

Reviewer 7: ```suggestion
- **Créer une nouvelle fixture ?** Multiplier les fixtures alourdi la modification du modèle de données.
```

Les fixtures présentent un autre inconvénient majeur : elles masquent des informations qui devraient être dans le test (cf principe [Expliciter le cas de test complet](#explicite-test-case))).

> Pour toutes ces raisons, il est préférable de privilégier la création de données à la volée à travers un DSL clair et paramétrable (ex: [Object Mother et Builder en Java](/posts/2019-01-11-object-mother-builder-java/)).

Reviewer 7: DSL ?

Author: En effet. J'ai remplacé par ça : "il est préférable de privilégier la création de données à la volée à travers des functions claires et paramétrables"

---
title: Les principes cachés du test
author: jb@sogilis.com & willy@sogilis.com

> date: 2020-10-01

Reviewer 5: :warning: changer la date si on ne veut pas que l'article apparaisse en 4ème position (après l'article sur VueX et les articles sur les drônes)

Author: On ferra ça juste avant de merger, oui.

image: /img/2020-10-01-principes-caches-du-test.jpg
altimage: '"Ten" by Mark Bonica is licensed with CC BY 2.0. To view a copy of this license, visit https://creativecommons.org/licenses/by/2.0/'
categories:
  - DÉVELOPPEMENT
tags:
  - test
---

## De quoi allons-nous parler ?

> Dans cet article, il est question de tests logiciel, mais pas n'importe quel test quel que soit le **contexte** :

Author: ```suggestion
Dans cet article, il est question de certains type de tests logiciels, ou plus précisément :
```

- 1 seul cas de test par 1 test
- Indépendants
- Rapides
- Automatisation du setup de dev (permettant l'exécution des tests en local)
- Executables sur toute machine de dev
- Exécutables plusieurs fois d'affilée

Dans cet article, nous allons parler de **stratégies peut-être moins connues** permettant de gagner encore un peu plus en lisibilité, maintenabilité, **utilisabilité au quotidien**. Ce ne sont **pas des règles absolues**, elles doivent être adaptées le cas échant, en fonction de la situation.

## 1. Réduire le boilerplate[^boilerplate]

L'idée derrière ce principe est de ne pas noyer le futur lecteur avec des détails inutiles dans le code du test[^testcode]. En effet, lorsqu'on est amené à travailler sur un test existant, il est préférable de **comprendre le cas de test rapidement**. Et si cette compréhension est **noyée au milieu de détails techniques**, cette tâche devient fastidieuse.

### Comment faire ?

Le boilerplate intervenant à plusieurs niveaux, les moyens sont donc multiples :

> - déplacer le code dans un "setup", "before each"…

Reviewer 5: ```suggestion
- déplacer le code du boilerplate dans des méthodes de facilitations de type "setup", "before each", "after"…
```

- externaliser le code dans une fonction dédiée.

L'inconvénient de la seconde solution, c'est qu'il reste encore du bruit dans le test (l'appel de la fonction), mais il n'est pas toujours possible de faire mieux.

> Concrètement, voilà ce que cela peut donner, en partant de ce test où la majorité du boilerplate est surligné :

Reviewer 5: ```suggestion
Voici un exemple de test, dans lequel le boilerplate (beaucoup trop gros) est surligné :
```

```go {linenos=inline,hl_lines=["3-12","19-25"]}
func TestHeroDAO_ResurrectAllKnights(t *testing.T) {
    // given
    sql.Register("mysql", &MySQLDriver{})
    dsn := fmt.Sprintf("%s:%s@tcp(%s)/%s", username, password, hostname, dbName)
    db, err := sql.Open("mysql", dsn)
    if err != nil {
        t.Logf("Error %s when opening DB\n", err)
        return
    }
    defer db.Close()

    db.Exec("TRUNCATE TABLE dead_hero")
    db.Exec("INSERT INTO dead_hero (name, allegiance, caste, dateOfDeath) VALUES ('Bohort', 'Gaunnes', 'knight', '25/10/1203')")

    // when
    kights := dao.NewHeroDAO(db).resurrectAllKnights()

    // then
    knightsContainsBohort := false
    for _, knight := range kights {
        if knight.name == "Bohort" {
            knightsContainsBohort = true
        }
    }
    if !knightsContainsBohort {
        t.Fatalf("Expected %s to contains 'Bohort'", knights)
    }
}
```

> En appliquant le principe a l'extrème, on peut arriver à ceci :

Reviewer 5: ```suggestion
En réduisant le boilerplate a l'extrème, on peut arriver à ceci :
```

```go {linenos=inline}
func TestHeroDAO_ResurrectAllKnights(t *testing.T, heroDAO *HeroDAO) {
    // given
    dbtest.InsertDeadKnight("Bohort")

    // when
    kights := heroDAO.resurrectAllKnights()

    // then
    assertOneKnightWithName(t, knights, "name")
}
```

- toute la partie connexion de la base de donnée a été déplacée en amont, pour être exécutée avant chaque test.
- idem avec le `TRUNCATE` (ligne 12).
- ainsi que la création de `HeroDAO` (ligne 16), tous les tests de ce fichier ayant besoin du même `HeroDAO`.

> De plus, ce boilerplate est souvent commun a plusieurs tests. Cette étape permet ainsi de le factoriser, facilitant sa maintenance en minimisant la duplication.

Reviewer 5: ```suggestion
De plus, notons que le boilerplate est souvent commun a plusieurs tests. Son extraction permet alors de le factoriser, facilitant sa maintenance, en minimisant la duplication. Attention toutefois à ne pas tomber dans l'excés en cherchant à factoriser trop de choses. L'élément important est d'extraire le boilerplate, sa factorisation n'est qu'une option, à étudier de façon opportuniste.
```

> Note : plus le test est haut niveau, plus ce boilerplate a tendance a être important.

Reviewer 5: ```suggestion
**Note :** nous avons pu observé que plus le test est haut niveau, plus ce boilerplate a tendance a être important.
```

Reviewer 5: Il faudrait dire ce que c'est « haut niveau » je pense non ? J'ai du mal  à trouver moi déjà.

## 3. Minimiser le jeu de données

Le jeu de données décrit les conditions initiales du test. Plus ces **données sont nombreuses**, plus il est **difficile de comprendre** ou **de debugger le test**.

C'est pourquoi, autant que faire se peut, il est préférable de limiter au maximum le jeu de test utilisé. Ainsi, les jeux de données communs a plusieurs tests (généralement utilisé avec les fixtures de test[^testfixture]) sont à éviter, mais nous en reparlerons plus tard.

## 4. Réfléchir au ROI

Lorsque l'on commence à écrire un test, cela vaut le coup de se poser 10 secondes pour réfléchir au [ROI](https://fr.wikipedia.org/wiki/Retour_sur_investissement) de ce test.

En effet, au-delà de ce qu'apporte un test, **il a aussi un coût**, qu'il soit **immédiat** (temps d'écriture du test) ou **récurrent** (maintenance, rallongement de la durée d'exécution des tests…). Tous ces facteurs ne sont pas à négliger car ils peuvent avoir des conséquences.

Exemple :

> Imaginez que je sois développeur sur une partie non critique de mon application, qu'il y ait beaucoup de couplage sur cette base de code et que, cherchant à produire un code de qualité, j'écrive des **tests unitaires pour chaque cas de chaque fonction**, y compris pour les fonctions les plus triviales. Résultat : à chaque petit refacto, beaucoup de tests doivent être adaptés, engendrant un **surcoût non négligeable**. Le temps passé à l'écriture de certains de ces tests aurait peut-être été plus utile sur une autre tâche (comme l'écriture de tests de plus haut niveau).

Reviewer 5: ```suggestion
Imaginez que je sois développeur sur une partie non critique de mon application, qu'il y ait beaucoup de couplage sur cette base de code et que, cherchant à produire un code de qualité, j'écrive des **tests unitaires pour chaque cas de chaque fonction**, y compris pour les fonctions les plus triviales. Résultat : chaque modification du code testé, fusse-t-elle un refacto de quelques lignes, risque d'engendrer des adaptations dans beaucoup de tests, ce qui implique un **coût non négligeable**. Il est alors important de se demander si le temps passé à l'écriture de certains de ces tests aurait peut-être été plus utile sur une autre tâche (comme l'écriture de tests de plus haut niveau).
```
